### DISCUSSION TITLE
Files in the project root

### DISCUSSION THREAD

**Discussion by nbolton** [deskflow/deskflow#7858] [Polls]
Blocks: #7850

Generally speaking, there are two approaches/philosophies to organizing a project root.

- **A:** Keep the project root clean
- **B:** Allow config files in the root

# Keep the project root clean

## Pros
- Improved readability: Easier to locate the main project files without sifting through numerous configs.
- Better organization: Config files can be grouped in dedicated folders (/config, etc.), making the structure more predictable.
- Clear separation: Keeps project code separate from setup and tooling specifics, which can aid in onboarding and maintainability.

## Cons
- Extra complexity: Requires additional effort to manage config paths or setup symlinks, especially if tools expect configs in the root.
- Tool compatibility: Not all tools easily support looking outside the root, which may mean extra configuration or workarounds.
- Maintenance overhead: The extra folders and paths need consistent naming and organization standards to avoid confusion over time.

# Allow config files in the root

## Pros
- Convenience: Easily accessible, with no need for extra paths; most tools work seamlessly with configs in the root.
- Standard practice: Many tools expect or default to root-level configs, so this setup aligns with common practices.
- Reduced setup: Simplifies project setup, as dependencies or team members don’t need to adjust or learn custom paths.

## Cons
- Cluttered workspace: Too many files in the root can make it hard to quickly find key files like README, LICENSE, or main scripts.
- Reduced focus: This can make it unclear which files are core to the project and which are for tooling or environment setup.
- Scalability issues: As more tools are added, config files accumulate, making the project harder to navigate.

# My take

I have observed that approach B seems more prevalent, for better or worse. I see so many repos now with a sea of config files in the root, and this may be even necessary for some projects... but where does it end? There always seems to be a new external/internal tool that needs a config file in the root and ours is filling up fast.

What if we try to push toward having a `config` dir instead...? Would that be fighting a losing battle? Keeping the root clean could offer more long-term benefits, especially since we want to bring in new contributors (i.e. make the project easier to grok). While it's a bit more upfront work, using a single config directory could save us from root clutter chaos. It does feel like we're fighting against the grain though...

**Comment by shymega** [comment]
With regards to Nix: this isn't something that can be voteable. I am happy to reduce the amount of config files - sure. But to move the necessary Nix files to a subdirectory is unergonomic, and would add complexity.

**Comment by sithlord48** [comment]
If we were going to put build files for any package in the root it would be for Debian, and I told them the same thing.  I do not want ANY packages info in the root of the project It just makes the root a mess.  Make a repo for the nix package the same as our debian maintainer will

**Comment by sithlord48** [comment]
The only package info that should be in the repo will be in our `/deploy` folder this will ONLY be items used to make our generic CPack packages. ANY other packages info does not belong in this repo.

**Comment by nbolton** [comment]
Closing this. The result seems to be that most devs don't mind config files in the root. I do agree that we should keep the root as clean as possible though.
